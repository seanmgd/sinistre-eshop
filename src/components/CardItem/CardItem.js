import { Link } from '@reach/router'
import PropTypes from 'prop-types'
import React from 'react'
import { ImgCard } from './CardItem.styles'

const SIZES = ['small', 'medium', 'large', 'x-large']
const COLORS = ['black', 'grey', 'dark-grey', 'light-grey']

const CardItem = ({
  to,
  imgUrl,
  label,
  labelSize,
  labelColor,
  hoverColor,
  ...rest
}) => (
  <Link to={to} {...rest}>
    <ImgCard
      imgUrl={imgUrl}
      hoverColor={hoverColor}
      label={label}
      labelSize={labelSize}
      labelColor={labelColor}
    />
  </Link>
)

CardItem.propTypes = {
  to: PropTypes.string.isRequired,
  labelColor: PropTypes.oneOf(COLORS).isRequired,
  labelSize: PropTypes.oneOf(SIZES).isRequired,
  hoverColor: PropTypes.oneOf(COLORS).isRequired,
  imgUrl: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
}

CardItem.defaultProps = {
  labelColor: 'grey',
  labelSize: 'small',
  hoverColor: 'black',
}

CardItem.SIZES = SIZES
CardItem.COLORS = COLORS

export default CardItem
