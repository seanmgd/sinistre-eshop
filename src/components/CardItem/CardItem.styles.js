import styled from 'styled-components'

const FONTSIZES = {
  small: '12px',
  medium: '16px',
  large: '20px',
  'x-large': '24px',
}

const getColor = (color, theme) => {
  const colorMapping = {
    black: theme.colors.black.base,
    grey: theme.colors.grey.base,
    'dark-grey': theme.colors.grey.dark,
    'light-grey': theme.colors.grey.light,
  }

  return colorMapping[color]
}

export const ImgCard = styled.div`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 600px;
  width: 30%;
  background-image: url(${({ imgUrl }) => imgUrl});
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;

  :hover {
    background-image: none;
    background-color: ${({ hoverColor, theme }) => getColor(hoverColor, theme)};
  }

  :after {
    content: '';
    color: ${({ labelColor, theme }) => getColor(labelColor, theme)};
    font-size: ${({ size }) => FONTSIZES[size]};
    font-weight: bold;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }

  :hover:after {
    content: ${({ label }) => `"${label}"`};
  }
`
