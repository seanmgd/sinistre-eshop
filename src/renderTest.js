import { render as renderTest } from '@testing-library/react'
import { AppProviders } from './AppProviders'
import theme from './styles/theme'

// Export it first to allow overrides.
export * from '@testing-library/react'
export { renderInAppProviders as render }

export function renderInAppProviders(node, options) {
  const { rerender, ...res } = renderTest(
    <AppProviders theme={theme} children={node} />,
    options,
  )

  return {
    ...res,
    rerender: node => rerender(<AppProviders theme={theme} children={node} />),
  }
}
