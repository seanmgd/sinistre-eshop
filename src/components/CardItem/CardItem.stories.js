import { files, select, text } from '@storybook/addon-knobs'
import React from 'react'
import CardItem from './CardItem'
import styled from 'styled-components'

export default {
  title: 'Data display/Card',
  component: CardItem,
}

const Container = styled.div`
  width: 30%;
`
export const ToStorybook = () => (
  <Container>
    <CardItem
      to="/"
      label={text('label', 'Demo label')}
      labelSize={select(
        'labelSize',
        CardItem.SIZES,
        CardItem.defaultProps.labelSize,
      )}
      labelColor={select(
        'labelColor',
        CardItem.COLORS,
        CardItem.defaultProps.labelColor,
      )}
      hoverColor={select(
        'hoverColor',
        CardItem.COLORS,
        CardItem.defaultProps.hoverColor,
      )}
      imgUrl={files('imgUrl', '.jpg, .png, .jpeg', '')}
    />
  </Container>
)

ToStorybook.story = {
  name: 'CardItem',
}
