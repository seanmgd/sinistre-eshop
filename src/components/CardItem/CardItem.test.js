import { fireEvent } from '@testing-library/react'
import { render } from '../../renderTest'
import { CardItem } from './CardItem'

function createProps(props = {}) {
  return { label: 'this is a test', to: '/', imgUrl: 'testimg', ...props }
}

describe('Card item', () => {
  it('should render card item', () => {
    const { container } = render(<CardItem {...createProps()} />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should spread other props on root element', () => {
    const testId = 'test-id'

    const props = createProps({
      onClick: jest.fn(),
      className: 'test-class',
      'data-testid': testId,
    })

    const { container, getByTestId } = render(<CardItem {...props} />)

    fireEvent.click(getByTestId(testId))

    expect(props.onClick).toHaveBeenCalledTimes(1)
    expect(container.firstChild).toMatchSnapshot()
  })

  CardItem.COLORS.forEach(color => {
    it(`should render a card item  with ${color} background color and text color`, () => {
      const props = createProps({
        hoverColor: color,
        labelColor: color,
      })

      const { container } = render(<CardItem {...props} />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  CardItem.SIZES.forEach(labelSize => {
    it(`should render a ${labelSize} text card item`, () => {
      const props = createProps({ labelSize })

      const { container } = render(<CardItem {...props} />)

      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
