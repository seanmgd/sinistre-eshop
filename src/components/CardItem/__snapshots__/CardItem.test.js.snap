// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card item should render a card item  with black background color and text color 1`] = `
<a
  aria-current="page"
  href="/"
>
  <div
    class="sc-bdfBwQ gwPNOT"
    label="this is a test"
  />
</a>
`;

exports[`Card item should render a card item  with dark-grey background color and text color 1`] = `
<a
  aria-current="page"
  href="/"
>
  <div
    class="sc-bdfBwQ boPxAL"
    label="this is a test"
  />
</a>
`;

exports[`Card item should render a card item  with grey background color and text color 1`] = `
<a
  aria-current="page"
  href="/"
>
  <div
    class="sc-bdfBwQ cBcSlb"
    label="this is a test"
  />
</a>
`;

exports[`Card item should render a card item  with light-grey background color and text color 1`] = `
<a
  aria-current="page"
  href="/"
>
  <div
    class="sc-bdfBwQ koGeHP"
    label="this is a test"
  />
</a>
`;

exports[`Card item should render a large text card item 1`] = `
<a
  aria-current="page"
  href="/"
>
  <div
    class="sc-bdfBwQ khjUgD"
    label="this is a test"
  />
</a>
`;

exports[`Card item should render a medium text card item 1`] = `
<a
  aria-current="page"
  href="/"
>
  <div
    class="sc-bdfBwQ khjUgD"
    label="this is a test"
  />
</a>
`;

exports[`Card item should render a small text card item 1`] = `
<a
  aria-current="page"
  href="/"
>
  <div
    class="sc-bdfBwQ khjUgD"
    label="this is a test"
  />
</a>
`;

exports[`Card item should render a x-large text card item 1`] = `
<a
  aria-current="page"
  href="/"
>
  <div
    class="sc-bdfBwQ khjUgD"
    label="this is a test"
  />
</a>
`;

exports[`Card item should render card item 1`] = `
<a
  aria-current="page"
  href="/"
>
  <div
    class="sc-bdfBwQ khjUgD"
    label="this is a test"
  />
</a>
`;

exports[`Card item should spread other props on root element 1`] = `
<a
  aria-current="page"
  class="test-class"
  data-testid="test-id"
  href="/"
>
  <div
    class="sc-bdfBwQ khjUgD"
    label="this is a test"
  />
</a>
`;
